src/string_with_vars.ml

```diff
@@ -104,105 +104,104 @@ let string_of_var syntax v =
+
+  let empty
+    : type a. a t -> a
+    = fun t ->
+      match t with
+      | Many -> []
+      | Single -> Value.String ""
 end
 
-module Expand_to(V: EXPANSION) = struct
-
-  let expand ctx t ~f =
-    match t.items with
-    | [Var (syntax, v)] when not t.quoted ->
-      (* Unquoted single var *)
-      (match f t.loc v with
-       | Some e -> Expand.Full.Expansion e
-       | None -> Expand.Full.String (string_of_var syntax v))
-    | _ ->
-      Expand.Full.String (List.map t.items ~f:(function
-        | Text s -> s
-        | Var (syntax, v) ->
-          match f t.loc v with
-          | Some x ->
-            if not t.quoted && V.is_multivalued x then
-              Loc.fail t.loc "please quote the string \
-                              containing the list variable %s"
-                (string_of_var syntax v)
-            else V.to_string ctx x
-          | None -> string_of_var syntax v)
-               |> String.concat ~sep:"")
-
-  let partial_expand ctx t ~f =
-    let commit_text acc_text acc =
-      let s = concat_rev acc_text in
-      if s = "" then acc else Text s :: acc
-    in
-    let rec loop acc_text acc items =
-      match items with
-      | [] -> begin
-          match acc with
-          | [] -> Expand.Partial.String (concat_rev acc_text)
-          | _  -> Unexpanded { t with items = List.rev (commit_text acc_text acc) }
-        end
-      | Text s :: items -> loop (s :: acc_text) acc items
-      | Var (syntax, v) as it :: items ->
-        match f t.loc v with
-        | None -> loop [] (it :: commit_text acc_text acc) items
-        | Some x ->
-          if not t.quoted && V.is_multivalued x then
-            Loc.fail t.loc "please quote the string containing the \
-                            list variable %s" (string_of_var syntax v)
-          else loop (V.to_string ctx x :: acc_text) acc items
-    in
-    match t.items with
-    | [Var (_, v)] when not t.quoted ->
-      (* Unquoted single var *)
-      (match f t.loc v with
-       | Some e -> Expand.Partial.Expansion e
-       | None   -> Expand.Partial.Unexpanded t)
-    | _ -> loop [] [] t.items
+module Partial = struct
+  type nonrec 'a t =
+    | Expanded of 'a
+    | Unexpanded of t
 end
 
-module String_expansion = struct
-  type t = string
-  let is_multivalued _ = false
-  type context = unit
-  let to_string () (s: string) = s
-end
-
-module S = Expand_to(String_expansion)
-
-let expand t ~f =
-  match S.expand () t ~f with
-  | Expand.Full.String s
-  | Expansion s -> s
-
-let partial_expand t ~f =
-  match S.partial_expand () t ~f with
-  | Expand.Partial.Expansion s -> Left s
-  | String s -> Left s
-  | Unexpanded s -> Right s
+let invalid_multivalue syntax ~var t x =
+  Loc.fail t.loc "Variable %s expands to %d values, \
+                  however a single value is expected here. \
+                  Please quote this atom."
+    (string_of_var syntax var) (List.length x)
+
+let expand_var syntax ~var ~dir ~f t =
```

Comment:
I think we can re-inline it now, this will avoid the intermediate option. BTW, I was finding the version with the `when not t.quoted` slightly easier to read

Reply:
Okay, sure I brought it back. I try to avoid guards because the exhaustivity checker was helpful to me here. But now that it is inline, it looks alright.